Add `permessage-deflate` support

kazk opened the pull request:
- Adds de/compression logic on top of the existing flow, requiring minimal changes. Shouldn't have any _major_ breaking changes.
- `client_max_window_bits`/`server_max_window_bits` are not supported at the moment because that requires `flate2/zlib` feature. We can enable them with a feature flag later if there's a demand.
- Passes autobahn except for 13.3 to 13.6, that requires supporting max window bits
- Can be used from crates like `warp` and `axum`.
  - `tokio-tungstenite`: https://github.com/snapview/tokio-tungstenite/pull/190
  - `warp`: https://github.com/seanmonstar/warp/compare/master...kazk:permessage-deflate
  - Example using `warp` with compression: https://github.com/qualified/lsp-ws-proxy/commit/89fce502cf792cfa89e814865bbe170666499c0d

This works for my use case, but I'd like some feedback from maintainers and users before polishing this to be merged.

- Does it make sense to think about supporting more PMCEs (e.g., `permessage-bzip2` and `permessage-snappy`)? These are not standardized as far as I know, and browsers don't support them.
- ~~I think including `flate2` without feature flag is fine (not large, defaults to pure Rust). What do you think? If there's a demand, we can add a flag for `flate2/zlib` and support max window bits.~~
- Should the default `WebSocketConfig` enable compression support?

---

- [x] `deflate` feature
- [x] Parsing `Sec-WebSocket-Extensions` more strictly with `headers` (https://github.com/kazk/tungstenite-rs/tree/permessage-deflate-rebased)
- [ ] Error on invalid negotiation response
- [ ] Error handling
- [ ] `deflate-zlib` feature for max window bits

---

Closes #2

application-developer-DA commented on src/extensions/compression/deflate.rs:
```diff
@@ -0,0 +1,289 @@
+        // * The negotiation offer contains multiple extension parameters with
+        //   the same name.
+        // * The server doesn't support the offered configuration.
+        'outer: for (_, offer) in
+            extensions::parse_header(extensions).iter().filter(|(k, _)| k == self.name())
```
Ok, I think I got the meaning of the function, but I believe that it could be simplified (and so also there would be no need for the `'outer`) if the function accepts the input for the deflate only, e.g. it would be something like:

```
fn generate_response(input_params: &[Param]) -> Response {
    let mut (param1, param2, ...) = (false, false);
    for param in params {
    }
    return Response::from(param1, param2, ...)
}
```

Does it make any sense?

application-developer-DA commented on src/extensions/compression/deflate.rs:
```diff
@@ -0,0 +1,289 @@
+        None
+    }
+
+    pub(crate) fn accept_response(&self, agreed: &[Param]) -> Result<DeflateContext, DeflateError> {
```
Oh, looks to something similar which I wrote in the comment above about `negotiation_response` :slightly_smiling_face:

application-developer-DA commented on src/extensions/compression/deflate.rs:
```diff
@@ -0,0 +1,289 @@
+                CLIENT_MAX_WINDOW_BITS => {}
+
+                _ => {
+                    //
```
An error? :)

application-developer-DA commented on src/extensions/compression/deflate.rs:
```diff
@@ -0,0 +1,289 @@
+            match self
+                .compressor
+                .compress_vec(&data[offset..], &mut output, FlushCompress::None)
+                .map_err(|e| DeflateError::Compress(e.into()))?
```
NB: What if we use the `#[from]` (from `thiserror`) on the error structure? Would it help to just use `?` after `compress_vec`? (e.g. spares us an explicit conversion with `map_err`)

kazk replied:
`compress_vec(...)?` only works if `DeflateError::Compress(flate2::CompressError)`.

I was thinking `Compress(std::io::Error)` and `Decompress(std::io::Error)` are better if we decide to support more PMCEs, and make them generic. Explicit conversions are required because both variants are `std::io::Error` (`#[from]` can't be used on both).

application-developer-DA commented on src/protocol/mod.rs:
```diff
@@ -348,8 +395,28 @@ impl WebSocketContext {
-            Message::Text(data) => Frame::message(data.into(), OpCode::Data(OpData::Text), true),
-            Message::Binary(data) => Frame::message(data, OpCode::Data(OpData::Binary), true),
+            Message::Text(data) => {
+                if let Some(pmce) = self.pmce.as_mut() {
+                    Frame::compressed_message(
+                        pmce.compress(data.as_bytes())?,
+                        OpCode::Data(OpData::Text),
+                        true,
+                    )
+                } else {
+                    Frame::message(data.into(), OpCode::Data(OpData::Text), true)
+                }
+            }
+            Message::Binary(data) => {
+                if let Some(pmce) = self.pmce.as_mut() {
+                    Frame::compressed_message(
+                        pmce.compress(&data)?,
+                        OpCode::Data(OpData::Binary),
+                        true,
+                    )
+                } else {
+                    Frame::message(data, OpCode::Data(OpData::Binary), true)
+                }
+            }
```
This looks ok now when we only have 2 choices, but I think ideally to make implementation more flexible, we probably want to have some function that we call from here, which then handles the implementation details (extensions, if any) inside its body. What do you think?

kazk replied:
Extracted to a separate method `fn prepare_data_frame(&mut self, data: Vec<u8>, opdata: OpData) -> Result<Frame>` that handles both text and binary.

application-developer-DA commented on src/protocol/mod.rs:
```diff
@@ -438,11 +505,16 @@ impl WebSocketContext {
             // value, the receiving endpoint MUST _Fail the WebSocket
             // Connection_.
+            let mut is_compressed = false;
             {
                 let hdr = frame.header();
-                if hdr.rsv1 || hdr.rsv2 || hdr.rsv3 {
+                if (hdr.rsv1 && self.pmce.is_none()) || hdr.rsv2 || hdr.rsv3 {
                     return Err(Error::Protocol(ProtocolError::NonZeroReservedBits));
                 }
+
+                if hdr.rsv1 && self.pmce.is_some() {
+                    is_compressed = true;
+                }
             }
```
I think we could use something like:

```
let is_compressed = {
... <that block of code> ...
}
```

So that we don't need a `mut`.

But a general question: what if we have more options? E.g. apart from `compressed` some other extensions? In this case we'll probably need to make the implementation a bit more generic and abstract, so that we don't clutter the code. Related to the previous comment.

kazk replied:
Made it immutable. We can also just use `rsv1` instead of `is_compressed` if you prefer.

If we have more extensions, we'll need to first check for any conflicts in their use of `rsv` bits during handshake. Then something like this:

```rust
let (is_compressed, use_e2, use_e3) = self.validate_header(frame.header())?;
// or
let (is_compressed, use_e2, use_e3) = {
    let hdr = frame.header();
    if (hdr.rsv1 && self.pmce.is_none()) || (hdr.rsv2 && self.ext2.is_none()) || (hdr.rsv3 && self.ext3.is_none()) {
        return Err(Error::Protocol(ProtocolError::NonZeroReservedBits));
    }
    (hdr.rsv1, hdr.rsv2, hdr.rsv3)
};
```

I don't know if an extension can use more than one bit.

application-developer-DA reviewed (commented):
Thanks for the change, that's quite a big addition :slightly_smiling_face: 

I tried to go through the code and also touch the questions that you asked in the PR description. Generally the implementation looks good and I really appreciate the effort! There are a couple of questions to answer / decisions to make that we need to decide upon though, I tried to comment on them close to the context in which they are introduced.

kazk commented:
@application-developer-DA

Thanks for reviewing! I'll start answering them soon. I opened this PR as a draft since it's not done (especially the client side like error handling, as you noticed), but wanted early feedback.

kazk commented on src/extensions/compression/deflate.rs:
```diff
@@ -0,0 +1,289 @@
+    pub client_no_context_takeover: bool,
+}
+
+impl Default for DeflateConfig {
+    fn default() -> Self {
+        Self {
+            compression: Compression::default(),
+            server_no_context_takeover: false,
+            client_no_context_takeover: false,
+        }
+    }
+}
```
Yeah, that works at the moment, but bools that defaults to `true` like `accept_server_no_context_takeover` seems like a common option, and we should probably support.

kazk commented on src/error.rs:
```diff
@@ -67,6 +70,9 @@ pub enum Error {
     /// HTTP format error.
     #[error("HTTP format error: {0}")]
     HttpFormat(#[from] http::Error),
+    /// Error from `permessage-deflate` extension.
+    #[error("deflate error: {0}")]
```
I forgot/missed that error messages are capitalized in `tungstenite`. I'll fix it.

By the way, it's recommended to be lowercased.

> The error message given by the `Display` representation of an error type should be lowercase without trailing punctuation, and typically concise.
>
> https://rust-lang.github.io/api-guidelines/interoperability.html#error-types-are-meaningful-and-well-behaved-c-good-err

kazk commented on src/extensions/mod.rs:
```diff
@@ -0,0 +1,81 @@
+    }
+}
+
+// NOTE This doesn't support quoted values
```
Yeah, I should've been more clear that this PR is not ready (draft), and/or left `TODO` comments instead. I mostly wanted feedback on the questions I had and overall approach, then work with the maintainers to discuss what's best for `tungstenite`.

For headers, are you open to using https://github.com/hyperium/headers ? I'm working on adding `SecWebsocketExtensions` there that provides an iterator on offers, so `tungstenite` doesn't need to do this.

application-developer-DA replied:
Sure, usage of `headers` is fine!

kazk replied:
I opened https://github.com/hyperium/headers/pull/88. Integrating with crates like `warp` and `axum` should be easy with that. But I think I need to update to hide more details for it to be merged.

kazk commented on src/extensions/compression/deflate.rs:
```diff
@@ -0,0 +1,289 @@
+
+    // This can be used for `WebSocket::from_raw_socket_with_compression`.
+    /// Returns negotiation response based on offers and `DeflateContext` to manage per message compression.
+    pub fn negotiation_response(&self, extensions: &str) -> Option<(HeaderValue, DeflateContext)> {
```
Yeah, like you wrote in the next comment, there might be more than one set of parameters for each extension, and this function should pick the first one that's acceptable.

I honestly don't like taking `&str` here, but I also wanted to minimize the number of public APIs (keep `extensions::parse_headers`, `Param`, etc. private) required for integration.

> a more generic implementation that could be re-used for other extensions if any.

Yeah, I've been thinking about this. Not just other PMCEs like I asked about, but also any other extensions. As far as I know, `permessage-deflate` is the only standardized extension, and I'm not sure if it's really worth implementing some generic framework for custom extensions.

kazk replied:
Moved parsing to `WebSocketConfig::accept_offers` which will pass only the offers to each extension.

kazk commented on src/extensions/compression/deflate.rs:
```diff
@@ -0,0 +1,289 @@
+    }
+
+    /// Value for `Sec-WebSocket-Extensions` request header.
+    pub(crate) fn negotiation_offers(&self) -> HeaderValue {
```
I actually had this named `generate_offers` at first. It's plural because I was thinking maybe we should support stacked offers.

The name "negotiation offers" was taken from the RFC, but I don't really like it. I'll change it to `generate_offer`.

kazk commented on src/extensions/mod.rs:
```diff
@@ -0,0 +1,81 @@
+                    params.push(param);
+                }
+            }
+            collected.push((Cow::from(name), params));
```
Yeah, that should be correct. I wasn't handling multiple `Sec-WebSocket-Extensions` request headers, but I just added it.

kazk commented on Cargo.toml:
```diff
@@ -28,6 +28,7 @@ __rustls-tls = ["rustls", "webpki"]
 base64 = "0.13.0"
 byteorder = "1.3.2"
 bytes = "1.0"
+flate2 = "1.0"
```
Yeah, a separate feature is usually my preference too.

> Requires a bit more coding to separate the things in the code / feature-gate them.

This was the main concern, and I was being lazy. I think it's starting to look easier to isolate now, so I'll try moving this behind a feature.

jenanwise commented:
@kazk Any chance this is still being worked on?

kazk commented:
I was planning to continue after getting https://github.com/hyperium/headers/pull/88 merged (see https://github.com/snapview/tungstenite-rs/pull/235#discussion_r742305530), but forgot. I've been using this branch with a patched `warp` in production.

Will try to take a look this week after work.

I think the following needs to be done to get this merged:

- Improve parsing `Sec-WebSocket-Extensions`. I did in https://github.com/hyperium/headers/pull/88, but we can include that here for now if necessary.
- Validate negotiation response
- Error handling in general

Nice to have:

- `deflate-zlib` feature to support max window bits

kazk commented:
For now, I've rebased and cleaned up the commits. https://github.com/kazk/tungstenite-rs/tree/permessage-deflate-rebased

I _might_ have some time over the weekend to continue.

kazk commented:
Yeah, I'd like to keep that in `headers`.

I'll update this PR to use `headers` (using my fork for now), and make sure it's useful for `tungstenite`. After making any necessary changes, I'll ask them to review that PR again and try to get that merged.

kazk commented:
I've updated to use `headers`. https://github.com/kazk/tungstenite-rs/commit/2e404a447c94eb05f3ee4059b7094bad06e20d1f

I'll try to spend some time during the week to finish it and open a new PR.

kazk commented:
@application-developer-DA

The autobahn results don't match the expected in https://github.com/snapview/tungstenite-rs/tree/master/ for both server and client. Is this a regression? I can open a separate PR to update if it just needs to be updated.

`diff expected master`:

```diff
     "7.7.10": {
       "behavior": "OK",
       "behaviorClose": "OK",
-      "remoteCloseCode": 1000,
+      "remoteCloseCode": 3000,
       "reportfile": "tungstenite_case_7_7_10.json"
     },
     "7.7.11": {
       "behavior": "OK",
       "behaviorClose": "OK",
-      "remoteCloseCode": 1000,
+      "remoteCloseCode": 3999,
       "reportfile": "tungstenite_case_7_7_11.json"
     },
     "7.7.12": {
       "behavior": "OK",
       "behaviorClose": "OK",
-      "remoteCloseCode": 1000,
+      "remoteCloseCode": 4000,
       "reportfile": "tungstenite_case_7_7_12.json"
     },
     "7.7.13": {
       "behavior": "OK",
       "behaviorClose": "OK",
-      "remoteCloseCode": 1000,
+      "remoteCloseCode": 4999,
       "reportfile": "tungstenite_case_7_7_13.json"
     },
     "7.7.2": {
       "behavior": "OK",
       "behaviorClose": "OK",
-      "remoteCloseCode": 1000,
+      "remoteCloseCode": 1001,
       "reportfile": "tungstenite_case_7_7_2.json"
     },
     "7.7.3": {
       "behavior": "OK",
       "behaviorClose": "OK",
-      "remoteCloseCode": 1000,
+      "remoteCloseCode": 1002,
       "reportfile": "tungstenite_case_7_7_3.json"
     },
     "7.7.4": {
       "behavior": "OK",
       "behaviorClose": "OK",
-      "remoteCloseCode": 1000,
+      "remoteCloseCode": 1003,
       "reportfile": "tungstenite_case_7_7_4.json"
     },
     "7.7.5": {
       "behavior": "OK",
       "behaviorClose": "OK",
-      "remoteCloseCode": 1000,
+      "remoteCloseCode": 1007,
       "reportfile": "tungstenite_case_7_7_5.json"
     },
     "7.7.6": {
       "behavior": "OK",
       "behaviorClose": "OK",
-      "remoteCloseCode": 1000,
+      "remoteCloseCode": 1008,
       "reportfile": "tungstenite_case_7_7_6.json"
     },
     "7.7.7": {
       "behavior": "OK",
       "behaviorClose": "OK",
-      "remoteCloseCode": 1000,
+      "remoteCloseCode": 1009,
       "reportfile": "tungstenite_case_7_7_7.json"
     },
     "7.7.8": {
       "behavior": "OK",
       "behaviorClose": "OK",
-      "remoteCloseCode": 1000,
+      "remoteCloseCode": 1010,
       "reportfile": "tungstenite_case_7_7_8.json"
     },
     "7.7.9": {
       "behavior": "OK",
       "behaviorClose": "OK",
-      "remoteCloseCode": 1000,
+      "remoteCloseCode": 1011,
       "reportfile": "tungstenite_case_7_7_9.json"
     }
```

kazk commented:
Probably changed by https://github.com/snapview/tungstenite-rs/pull/246

application-developer-DA commented:
@kazk thanks for the investigation! The corresponding PRs have been merged.

kazk commented:
Superseded by #328. If anyone is using this branch, I'm not deleting it because I also have some projects depending on it.
